src/kalshi.rs

```diff
@@ -0,0 +1,377 @@
+        let mut return_string = self.event.underlying.clone();
+        // For each variable in market.rulebook_variables, substitute ||variable|| with the value of the variable
+        let rulebook = self.get_market().rulebook_variables.clone();
+        // For each key in rulebook
+        for key in rulebook.as_object().unwrap().keys() {
+            // Remove the surrounding "" marks
+            let mut replacement_value = rulebook[key].to_string();
+            replacement_value = replacement_value[1..replacement_value.len()-1].to_string();
+
+            // We've seen instances with and without spaces
+            return_string = return_string.replace(&format!("||{}||", key), &replacement_value);
+            return_string = return_string.replace(&format!("|| {} ||", key), &replacement_value);
+        }
+        return return_string
+
+    }
+
+    pub fn get_resolution_sources_markdown(&self) -> String {
+        let sources = self
+            .event
+            .settlement_sources
+            .iter()
+            .map(|source| format!("<{}>", source.url.clone()))
+            .collect::<Vec<String>>();
+
+        // Return "" if there are none
+        if sources.is_empty() {
+            return "".to_string();
+        }
+        format!("\n\n\n**Resolution sources**\n\n{}", sources.join(", "))
+    }
+
+    pub fn get_binary_resolution(&self) -> Result<Option<BinaryResolution>> {
+        if self.is_resolved() {
+            match self.get_market().result {
+                Some(KalshiResult::Yes) => Ok(Some(BinaryResolution::Yes)),
+                Some(KalshiResult::No) => Ok(Some(BinaryResolution::No)),
+                Some(KalshiResult::StillOpen) => Ok(None),
+                None => Ok(None),
+            }
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+impl TryInto<Question> for &KalshiQuestion {
+    type Error = anyhow::Error;
+
+    fn try_into(self) -> Result<Question> {
+        Ok(Question {
+            source: QuestionSource::Kalshi,
+            source_url: self.full_url(),
+            source_id: self.id.clone(),
+            question: self.get_market().title.clone(),
+            criteria: Some(self.get_criteria_and_sources()),
+            end_date: self.get_market().expiration_date,
+        })
+    }
+}
+
+#[derive(Deserialize, Debug, Clone)]
+pub struct KalshiQuestionsResponse {
+    pub events: Vec<Event>,
+}
+
+#[derive(Deserialize, Debug, Clone)]
+pub struct KalshiQuestion {
+    pub event: Event,
+    #[serde(skip)]
+    pub id: String,
+}
+
+#[derive(Deserialize, Debug, Clone)]
+#[serde(rename_all = "snake_case")]
+pub struct Event {
+    pub series_ticker: String,
+    pub ticker: String,
+    pub markets: Vec<Market>,
+    pub settlement_sources: Vec<SettlementSource>,
+    pub underlying: String,
+}
+
+#[derive(Deserialize, Debug, Clone)]
+pub struct SettlementSource {
+    pub name: String,
+    pub url: String,
+}
+
+#[derive(Deserialize, Debug, Clone)]
+pub struct Market {
```

Comment:
It looks like Kalshi has scalar markets as well. Are we parsing that anywhere? We don't want to mirror those as if they're binary.

Reply:
I'm not seeing any market types that aren't binary. Could you share a link?

Reply:
Maybe they don't actually exist? They're referred to in the docs for v2 (`market_type` can be `scalar` or `binary`).

Reply:
Something to watch out for in the v2 future.